### ISSUE TITLE
Docker driver fails to pull complete .tar file via artifact stanza from private repo

### ISSUE THREAD

**Issue by ctmiller8** [hashicorp/nomad#16251]
### Nomad version
Output from `nomad version`
Nomad v1.5.0-beta.1 (3d735e74efc62440817996162d2edff32e66095b)

### Operating system and Environment details
3 Server Cluster (Linux VMs: ubuntu 20.04.5)

6 Clients (2 Win10 VMs 10.0.19045.2486 Build 19045.2486/4 Linux VMs ubuntu 20.04.5 )

Docker Driver on Linux Clients running Docker Engine with below details:
Client: Docker Engine - Community
 Version:           23.0.1
 API version:       1.42
 Go version:        go1.19.5

### Issue
When attempting to pull a .tar file from a private repo via the artifact stanza, the target Linux client writes the below error messages to the client logs and fails to download the complete tarball and the job fails:

[INFO]  client.alloc_runner.task_runner: Task event: alloc_id=31e6219c-364c-2af0-695f-28883dc1f39f task=controller-web type="Driver Failure" msg="Error processing tar file(exit status 1): unexpected EOF" failed=false

Nomad Linux client also adds the below confirmation in client logs that artifact was downloaded successfully:

[DEBUG] client.artifact: sub-process: OUTPUT="artifact download was a success"

However, while monitoring the .tar file download size via the allocation's /local fs, the file download stops short of the correct total .tar file size during each allocation attempt:

.tar actual file size = 1.33 GB
.tar incomplete dl file size = 272.75 MiB

I've attempted to add additional artifact stanza settings on the impacted clients per the below with no success:

       artifact = {
        http_read_timeout = "0"
        http_max_size = "0"          
        }
       }

Any help is greatly appreciated - thanks in advance for your support.

### Reproduction steps
1. Run a job with artifact stanza to pull a tarball from a private repo (target tarball is ~1.33 GB)
2. Nomad job stops short of completing the full download of the tarball (incomplete tarball size is ~272.75 MiB) and throws EOF when attempting docker load

#### Expected Result
1. Job is started successfully after tarball download via artifact stanza has completed.

#### Actual Result
1. Job fails to start successfully as artifact download does not complete and docker load throws EOF error messages.

### Job file (if appropriate)

job "controller-web-02" {
   namespace = "npe-fusion"
   type = "service"
   datacenters = ["dev"]

  constraint {
   attribute = "${node.unique.name}"
   value     = "somenode"
}

  group "controller-web" {
    count = 1

    network {
      port "http" {
        static = 8082
      }
    }

    service {
      name = "controller-web"
      provider = "nomad" 

      check {
        name     = "alive"
        type     = "tcp"
        port     = "http"
        interval = "10s"
        timeout  = "2s"
      }
    }

    task "controller-web" {
      driver = "docker"

      artifact {
        source = "https://private.repo.com/stash/users/repos/rpa/raw/controller_web.tar"

        options {
          archive = false
        }
    }

      config {
        load = "controller_web.tar"
        image = "controller_web:latest"

        ports = ["http"]

        mount = {
          type = "volume"
          target = "/home/app/web/staticfiles"
          source = "static_volume"
          readonly = "false"
        }

      resources {
        cpu    = 1000
        memory = 1000
       }
       
      }
    }
  }
}

### Nomad Server logs (if appropriate)
N/A

### Nomad Client logs (if appropriate)
*REDACTED*

[INFO]  client.alloc_runner.task_runner: Task event: alloc_id=e1256d2f-e115-5b68-d10e-d34b233caeda task=controller-web type="Downloading Artifacts" msg="Client is downloading artifacts" failed=false

[DEBUG] client.alloc_runner.task_runner.task_hook.artifacts: downloading artifact: alloc_id=e1256d2f-e115-5b68-d10e-d34b233caeda task=controller-web artifact="https://private.repo.com/stash/users/repos/rpa/raw/controller_web.tar" aid="FWEevT3I9/grf+NBn0704AG09YTyNkllPKusNJdPfxo"

[DEBUG] client.artifact: get: source="https://private.repo.com/stash/users/repos/rpa/raw/controller_web.tar" destination=local/

[DEBUG] client.artifact: sub-process: OUTPUT="artifact download was a success"

[DEBUG] client.driver_mgr.docker: loading image from disk: driver=docker archive=/nomad/data/alloc/31e6219c-364c-2af0-695f-28883dc1f39f/controller-web/local/controller_web.tar

[INFO]  client.alloc_runner.task_runner: Task event: alloc_id=31e6219c-364c-2af0-695f-28883dc1f39f task=controller-web type="Driver Failure" msg="Error processing tar file(exit status 1): unexpected EOF" failed=false

[ERROR] client.alloc_runner.task_runner: running driver failed: alloc_id=31e6219c-364c-2af0-695f-28883dc1f39f task=controller-web error="Error processing tar file(exit status 1): unexpected EOF"

[INFO]  client.alloc_runner.task_runner: not restarting task: alloc_id=31e6219c-364c-2af0-695f-28883dc1f39f task=controller-web reason="Error was unrecoverable"

[INFO]  client.alloc_runner.task_runner: Task event: alloc_id=31e6219c-364c-2af0-695f-28883dc1f39f task=controller-web type="Not Restarting" msg="Error was unrecoverable" failed=true

**Comment by shoenig** [comment]
Hi @ctmiller8 thanks for the report - the first thing I would suggest here is setting the `checksum` in the `artifact` block, so that `go-getter` can confirm what it downloaded is what you expect.

go-getter cheksum docs: https://github.com/hashicorp/go-getter#checksumming
setting checksum in nomad docs: https://developer.hashicorp.com/nomad/docs/job-specification/artifact#download-and-verify-checksums

**Comment by ctmiller8** [comment]
Hi @shoenig thanks for the quick reply.

I followed your suggestion, added checksum details to my artifact stanza and re-ran the job (multiple times on different clients) but the complete tarball still fails to download and checksums are not matching.

Please see the below error messages:

[ERROR] client.artifact: sub-process: OUTPUT="failed to download artifact: Checksums did not match for /nomad/data/alloc/4780590e-b2d3-dd9c-03f0-f23c84a94690/controller-web/local/controller_web.tar."

[ERROR] client.artifact: sub-process: OUTPUT="Expected: e3a9bd7fbd7cc96d41b4bd8a7a1c4848"

[ERROR] client.artifact: sub-process: OUTPUT="Got: 261d17f234ea9bdfbcc6b03155ed2aa1"

[ERROR] client.artifact: sub-process: OUTPUT="*md5.digest"

[INFO]  client.alloc_runner.task_runner: Task event: alloc_id=7f53778f-af1b-6d1d-c556-85c5bcc79894 task=controller-web type="Failed Artifact Download" msg="failed to download artifact \"https://private.repo.com/stash/users/repos/rpa/raw/controller_web.tar\": getter subprocess failed: exit status 1" failed=false

**Comment by ctmiller8** [comment]
Hi @shoenig  I believe I've found a solution to my specific problem.

In short, my private repo was not configured to support pushing/pulling large files - I followed the basic enable [bitbucket LFS steps](https://support.atlassian.com/bitbucket-cloud/docs/use-git-lfs-with-bitbucket/) .

The key takeaways for my specific issue were (1) adding the file filter for all .tar files in my repo and (2) re-committing and re-pushing all of the .tar files with LFS enabled.

Hopefully this info can help someone else in the future.
